tests/integrated/test_suite

```diff
@@ -58,6 +62,100 @@ try:
 except RuntimeError:
     raise RuntimeError("The all-tests flag needs to be set by passing --all rather then --set-bool all_tests=True")
 
+##################################################################
+# Parallel stuff
+
+# Check what we are supposed to be doing:
+if 'MAKEFLAGS' in os.environ:
+    mf=os.environ['MAKEFLAGS']
+    for token in mf:
+        if token == "n":
+            # Print recipies:
+            # We ignore this for now ...
+            sys.exit(0)
+        elif token == "q":
+            # Question mode:
+            # We are never up-to-date
+            sys.exit(1)
+        elif token in "ksBd":
+            # k - keep going
+            # s - silent
+            # B - always build (we do)
+            # d - debug
+            pass
+        elif token == " ":
+            break
+        else:
+            # Not implemented
+            print("mode '%s' not implemented"%mf[0])
+            #sys.exit(42)
+else:
+    mf = ""
+
+num_jobs=1
+if args.jobs:
+    num_jobs = args.jobs[0]
+js_read=None
+if "--jobserver-auth=" in mf:
+    # running in parallel
+    offset=mf.index("--jobserver-auth=")
+    file_desc=[int(i) for i in re.findall("\d+",mf[offset:])[0:2]]
+    try:
+        js_read = open(file_desc[0], "r")
+        js_write = open(file_desc[1], "w")
+    except:
+        print("Warning - we are not part of the job pool 😭, running in serial")
+        file_desc = None
+        js_read = None
+    if js_read is not None:
+        js_tokens = js_read.read()
+        num_jobs = len(js_tokens)+1
+
+class Test(threading.Thread):
+    def __init__(self, name, cmd):
+        threading.Thread.__init__(self)
+        self.name = name
+        self.cmd = cmd
+        self.l = threading.local()
+        self.cost = self._cost()
+    def run(self):
+        # Check requirements
+        self.l.req_met, self.l.req_expr = requirements.check(self.name+"/runtest")
+        self.l.start_time = time.time()
+        if not self.l.req_met:
+            print(output % self.name +
+                  "S - {0} => False".format(self.l.req_expr))
+            sys.stdout.flush()
+            self.status = 0
+        else:
+            # Run test, piping stdout so it is not sent to console
+            self.status, self.output = shell("cd "+self.name+";"+self.cmd, pipe=True)
+            print(output % self.name, end='')
+            if self.status != 0:
+                # ❌ Failed
+                print(u"\u274C", end='')  # No newline
+            else:
+                # ✓ Passed
+                print(u"\u2713", end='')  # No newline
+            print(" %7.3f s" % (time.time() - self.l.start_time))
+            sys.stdout.flush()  # Make sure '✓' or '❌' is flushed
+    def _cost(self):
+        if self.cmd == "make":
+            return 1
+        with open(self.name+"/"+self.cmd,"r") as filein:
+            contents = filein.read()
+            # Find all lines starting with '#requires' or '#Requires:'
```

Comment:
> Parallel builds work for me on Marconi. Definite improvement +1

Would you mind sharing instructions on how to run recent BOUT++ on marconi?

> It seems as though the tests still run consecutively though, and it looks like that is not the intention (`#cores:8`, etc.)? If I do `make -j 32 check-integrated-tests`, the sum of the times for running the individual tests is equal to the total run time. I haven't tried to dig into why...

The tests should run in parallel. That sounds like a bug if the total runtime equals the sum of the runs, especially on 32 cores. I'll investigate ...

> When I run `make -j 32 check` the unit tests, integrated tests and MMS tests all start building at the same time, and their output overlaps. That's what always used to happen - just wondering if the jobserver takes that into account, or might it be more efficient to run something like `make -j 32 check-unit-tests && make -j 32 check-integrated-tests && make -j 32 check-mms-tests`?

Yes, that it is not nice. I have been wondering whether it would be nicer to implement the job server on a higher level, so it can run all tests in parallel. That would produce a nicer output. Also the user probably doesn't care which tests are run ...

Running one after another is what I thought about, that could be achieved by having the different targets sequential in a recipe. However, the unit tests run in serial, so it makes sense to run them in parallel if you have many cores. In the case of few cores (e.g. travis) not, as then we might schedule some expensive (i.e. tests requiring many cores) while other things are running, and it would be better to wait until the other tests are done. However in that case every test_suite gets only one job, so it doesn't know about the other tests. Moving the script to `tests/` and schedule all jobs would help in that case ...